--- Qualitative_ott.orig	2022-09-11 21:25:08.526629093 -0400
+++ Qualitative_ott.v	2022-09-11 22:23:41.464430761 -0400
@@ -650,19 +650,32 @@
      Typing W  (q_join  psi0   psi )  a A ->
      Typing W psi b  (open_tm_wrt_tm  B   a )  ->
      Typing W psi (a_WPair a psi0 b) (a_WSigma psi0 A B)
- | T_LetPairIrrel : forall (L:vars) (W:context) (psi psi0:grade) (a c C B:tm) (s:sort) (A:tm) (phi:grade),
+ | T_LetPairIrrel : forall (L:vars) (W:context) (psi psi0:grade) (a c C B A:tm) (s:sort),
       (  q_C   <  psi0 )  ->
       (  q_C   <  psi )  ->
-      ( forall x , x \notin  L  -> Typing W  q_Top   ( (a_Pi psi  ( open_tm_wrt_tm B (a_Var_f x) )  C) )  (a_Type s) )  ->
+      ( forall x , x \notin  L  -> Typing ((x ~ (q_Top, a_WSigma psi0 A B)) ++ W)  q_Top
+                                (open_tm_wrt_tm C (a_Var_f x))  (a_Type s))  ->
      Typing W psi a  ( (a_WSigma psi0 A B) )  ->
-      ( forall x , x \notin  L  -> Typing   (  ( x ~( psi0 , A ))  ++  (  (meet_ctx_l   q_C    W )  )  )    q_C   ( open_tm_wrt_tm c (a_Var_f x) )   ( (a_Pi phi  ( open_tm_wrt_tm B (a_Var_f x) )  C) )  )  ->
-      ( forall y , y \notin  L  -> Typing W psi (a_LetPair psi0 a c)  ( open_tm_wrt_tm C (a_Var_f y) )  ) 
- | T_LetPair : forall (L:vars) (W:context) (psi psi0:grade) (a c C B:tm) (s:sort) (A:tm) (phi:grade),
+     ( forall x , x \notin  L  ->
+             forall y,  y \notin L \u {{x}} ->
+              Typing   (  ( x ~( q_join psi0 psi , A ))  ++  (  (meet_ctx_l   q_C    W )  )  )    q_C
+                ( open_tm_wrt_tm c (a_Var_f x) )
+                (a_Pi q_Bot (open_tm_wrt_tm B (a_Var_f x))
+                   (close_tm_wrt_tm y (open_tm_wrt_tm C (a_WPair (a_Var_f x) psi0 (a_Var_f y))))))  ->
+      ( Typing W psi (a_LetPair psi0 a c)  ( open_tm_wrt_tm C a )  ) 
+ | T_LetPair : forall (L:vars) (W:context) (psi psi0:grade) (a c C B A:tm) (s:sort),
        (  ( psi  <=   q_C  )  )   \/   (  ( psi0  <=   q_C  )  )   ->
-      ( forall x , x \notin  L  -> Typing W  q_Top   ( (a_Pi psi  ( open_tm_wrt_tm B (a_Var_f x) )  C) )  (a_Type s) )  ->
+       ( forall x , x \notin  L  -> Typing ((x ~ (q_Top, a_WSigma psi0 A B)) ++ W)  q_Top
+                                 (open_tm_wrt_tm C (a_Var_f x))  (a_Type s))  ->
+       
      Typing W psi a  ( (a_WSigma psi0 A B) )  ->
-      ( forall x , x \notin  L  -> Typing   (  ( x ~(  (q_join  psi0   psi )  , A ))  ++ W )   psi  ( open_tm_wrt_tm c (a_Var_f x) )   ( (a_Pi phi  ( open_tm_wrt_tm B (a_Var_f x) )  C) )  )  ->
-      ( forall y , y \notin  L  -> Typing W psi (a_LetPair psi0 a c)  ( open_tm_wrt_tm C (a_Var_f y) )  ) 
+     ( forall x , x \notin  L  ->
+             forall y,  y \notin L \u {{x}} ->
+                   Typing   (  ( x ~(  (q_join  psi0   psi )  , A ))  ++ W )   psi
+                     ( open_tm_wrt_tm c (a_Var_f x) )
+                       (a_Pi q_Bot (open_tm_wrt_tm B (a_Var_f x))
+                          (close_tm_wrt_tm y (open_tm_wrt_tm C (a_WPair (a_Var_f x) psi0 (a_Var_f y))))))  ->
+      ( Typing W psi (a_LetPair psi0 a c)  ( open_tm_wrt_tm C a )  ) 
  | T_SSigma : forall (L:vars) (W:context) (psi psi0:grade) (A B:tm) (s3 s1 s2:sort),
      Typing W psi A (a_Type s1) ->
       ( forall x , x \notin  L  -> Typing  (  ( x ~( psi , A ))  ++ W )  psi  ( open_tm_wrt_tm B (a_Var_f x) )  (a_Type s2) )  ->
@@ -692,8 +705,9 @@
      Typing W psi a2 A2 ->
      Typing W  q_Top  (a_Sum A1 A2) (a_Type s) ->
      Typing W psi (a_Inj2 a2) (a_Sum A1 A2)
- | T_Case : forall (L:vars) (W:context) (psi psi0:grade) (a b1 b2 B:tm) (z:tmvar) (A1 A2:tm) (s:sort) (B1 B2:tm),
-     Typing  (  (  ( z ~(  q_C  , (a_Sum A1 A2) ))  ++ W )  )   q_Top  B (a_Type s) ->
+ | T_Case : forall (L:vars) (W:context) (psi psi0:grade) (a b1 b2 B:tm) (A1 A2 B1 B2:tm)  (s:sort),
+     (forall z, z \notin L -> Typing  (  (  ( z ~(  q_C  , (a_Sum A1 A2) ))  ++ W )  )   q_Top
+                           (open_tm_wrt_tm B (a_Var_f z)) (a_Type s)) ->
      Typing W psi a (a_Sum A1 A2) ->
       ( forall x , x \notin  L  ->   ( open_tm_wrt_tm B1 (a_Var_f x) )  =   (open_tm_wrt_tm  B   (a_Inj1 (a_Var_f x)) )    )  ->
       ( forall y , y \notin  L  ->   ( open_tm_wrt_tm B2 (a_Var_f y) )  =   (open_tm_wrt_tm  B   (a_Inj2 (a_Var_f y)) )    )  ->
